options/internal/include/bits/ensure.h

```diff
@@ -21,6 +21,15 @@ void __ensure_warn(const char *assertion, const char *file, unsigned int line,
 #define MLIBC_MISSING_SYSDEP() __ensure_warn("Library function fails due to missing sysdep", \
 		__FILE__, __LINE__, __func__)
 
+#define MLIBC_CHECK_OR_ENOSYS(sysdep) ({ \
+			if (!(sysdep)) { \
+				__ensure_warn("Library function fails due to missing sysdep", \
+					__FILE__, __LINE__, __func__); \
+				errno = ENOSYS; \
+				return -1; \
+			} \
```

Comment:
Return `sysdep` here.

Reply:
We'd need to be passed the arguments then, wouldn't we

Reply:
Ah no you don't mean to call it, just return the function pointer like you wrote above. I'm not sure I agree with that tbh. I think the macro is okay how it is now.

Reply:
Well, consider this:

```cpp
ssize_t listxattr(const char *path, char *list, size_t size) {
	MLIBC_CHECK_OR_ENOSYS(mlibc::sys_listxattr);

	ssize_t nread;
	if (int e = mlibc::sys_listxattr(path, list, size, &nread); e) {
		errno = e;
		return -1;
	}
	return nread;
}

ssize_t llistxattr(const char *path, char *list, size_t size) {
	MLIBC_CHECK_OR_ENOSYS(mlibc::sys_llistxattr);

	ssize_t nread;
	if (int e = mlibc::sys_listxattr(path, list, size, &nread); e) {
		errno = e;
		return -1;
	}
	return nread;
}

ssize_t flistxattr(int fd, char *list, size_t size) {
	MLIBC_CHECK_OR_ENOSYS(mlibc::sys_flistxattr);

	ssize_t nread;
	if (int e = mlibc::sys_flistxattr(fd, list, size, &nread); e) {
		errno = e;
		return -1;
	}
	return nread;
}
```

I hope a realistic scenario like this (this happened while I was implementing these) demonstrates my concern. This subtle of an error would've been missed had I merely not also been writing tests that target edge cases (the difference in the two confused functions is that small).

Reply:
I'll let korona decide

Reply:
I think I'll vote for the solution as it currently is, mostly because returning the sysdep wouldn't allow for code such as ``MLIBC_CHECK_OR_ENOSYS(mlibc::sys_getpid && mlibc::sys_kill, -1);``, which is a useful usecase. Unless there are some structured binding memes involved...

Reply:
For the structured binding idea we could simply have the macro call into a C++ function that generates the structured binding right?

You make a good point on the code duplication, and I understand the concern raise w.r.t. bug resilience.

Reply:
For example, if I look at glibc, it's quite difficult to understand how the syscall entry points are generated unless you spend some hours to read the code. This should definitely be avoided.

Reply:
The number of sysdep wrappers is bounded so a bit of code duplication does not hurt. Plus, there are subtle differences. For example, compare `madvise()` (which uses errno) vs. `posix_madvise()` (which returns an error code).

Reply:
I'll take it over barely detectable (even at runtime) bugs. This behaviour isn't that unpredictable if you look at the call site of such a return either.

Reply:
IMO the code readability is more important than the small code writabiloty benefit here.